micro-rdk/src/esp32/entry.rs

```diff
@@ -159,146 +177,204 @@ pub async fn serve_web_inner(
+    <S as WifiCredentialStorage>::Error: Sync + Send + 'static,
 {
-    use super::conn::mdns::Esp32Mdns;
-    use crate::common::conn::mdns::Mdns;
-    use std::ffi::CString;
-    let _ = Timer::after(std::time::Duration::from_millis(150)).await;
-    let hostname = format!(
-        "provisioning-{}-{}",
-        info.get_model(),
-        info.get_manufacturer()
-    );
-    let mut mdns = Esp32Mdns::new(hostname)?;
-
-    let srv = ProvisioningServiceBuilder::new().with_provisioning_info(info);
-
-    let srv = if let Some(error) = last_error {
-        srv.with_last_error(error)
-    } else {
-        srv
+    use crate::{
+        common::provisioning::server::serve_provisioning_async,
+        esp32::{
+            conn::mdns::Esp32Mdns, provisioning::wifi_provisioning::Esp32WifiProvisioningBuilder,
+        },
     };
-    let srv = srv.build(storage.clone());
-    let listen = TcpListener::bind(ip.to_string() + ":0")?;
-    let listen: Async<TcpListener> = listen.try_into()?;
-
-    let port = listen.get_ref().local_addr()?.port();
-
-    mdns.add_service(
-        "provisioning",
-        "_rpc",
-        "_tcp",
-        port,
-        &[("provisioning", "")],
-    )?;
-    loop {
-        let incoming = listen.accept().await;
-        let (stream, _) = incoming?;
-
-        // The provisioning server is exposed over unencrypted HTTP2
-        let stream = Esp32Stream::LocalPlain(stream);
-
-        let ret = ProvisoningServer::new(srv.clone(), exec.clone(), stream).await;
-        if ret.is_ok() && storage.has_stored_credentials() {
-            let creds = storage.get_robot_credentials().unwrap();
-            let app_config = AppClientConfig::new(
-                creds.robot_secret().to_owned(),
-                creds.robot_id().to_owned(),
-                "".to_owned(),
-            );
-
-            let conn =
-                Esp32Stream::TLSStream(Box::new(Esp32TLS::new_client().open_ssl_context(None)?));
-            let client = GrpcClient::new(conn, exec.clone(), "https://app.viam.com:443").await?;
-            let builder = AppClientBuilder::new(Box::new(client), app_config.clone());
-
-            let mut client = builder.build().await?;
-            let certs = client.get_certificates().await?;
-
-            let serv_key = CString::new(certs.tls_private_key).unwrap();
-            let serv_key_len = serv_key.as_bytes_with_nul().len() as u32;
-            let serv_key: *const u8 = serv_key.into_raw() as *const u8;
-
-            let tls_certs = CString::new(certs.tls_certificate)
-                .unwrap()
-                .into_bytes_with_nul();
-
-            return Ok((
-                app_config,
-                Esp32TLSServerConfig::new(tls_certs, serv_key, serv_key_len),
-            ));
+
+    use super::conn::network::Esp32WifiNetwork;
+
+    let mut last_error: Option<Box<dyn std::error::Error>> = None;
+
+    let info = info.unwrap_or_default();
+
+    let network = loop {
```

Comment:
if we made the network and executor types generic parameters, could we actually put this function in common? Edit: In fact, I have this question about all the functions in this file

Reply:
It's possible but I think it would be best to write as a macro in the mean time, We can do it in a later PR. A full generic approach is challenging because we have many type that are very similar but slightly different. In the future we can start an effort to de-duplicate code between native and esp32 but out of scope for this PR

Reply:
Yeah, fair enough